### ISSUE TITLE
no-unused-state false positive with UNSAFE_componentWillUpdate nextState

### ISSUE THREAD

**Issue by secondfry** [yannickcr/eslint-plugin-react#2591]
Seems to be related to #2209.

```js
import { Component } from 'react';

export default class MyComponent extends Component {
    constructor(props) {
        super(props);

        this.setState({
            a: 'q'
        });
    }

    UNSAFE_componentWillUpdate(_, nextState) {
        this.someHelper(nextState);
    }

    someHelper(state = this.state) {
        const { a } = state;
        return a;
    }
}
```
```
{
    "parserOptions": {
        "sourceType": "module"
    },
    "extends": [
        "eslint:recommended",
        "plugin:react/recommended"
    ],
    "rules": {
        "react/no-unused-state": "error"
    },
    "env": {
        "es6": true
    }
}
```
```
{
  "devDependencies": {
    "eslint": "^6.8.0",
    "eslint-plugin-react": "^7.19.0"
  },
  "dependencies": {
    "react": "^16.13.0"
  }
}
```

This will trigger `  8:13  error  Unused state field: 'a'  react/no-unused-state` even tho `this.state.a` was destructurized in `someHelper` method and used.

**Comment by secondfry** [comment]
Same with:
```js
import { Component } from "react";

export default class MyComponent extends Component {
  constructor(props) {
    super(props);

    this.setState({
      a: "q"
    });
  }

  UNSAFE_componentWillUpdate(_, { a }) {
    this.someHelper({ a });
  }

  someHelper({ a } = this.state) {
    return a;
  }
}
```

Workaround with:
```
import { Component } from "react";

export default class MyComponent extends Component {
  constructor(props) {
    super(props);

    this.setState({
      a: "q"
    });
  }

  UNSAFE_componentWillUpdate(_, { a }) {
    this.someHelper(a);
  }

  someHelper(a = this.state.a) {
    return a;
  }
}
```
but if you have more than 1 parameter it seems to be bad (DRY?).

**Comment by ljharb** [comment]
In general, you shouldn't be passing around state or props objects anyways. Always explicitly destructure them as soon as you get them, and pass those values around to wherever you need them.
